<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#-Homework-1-(Topics-1-4:-Data-Structures,-Indexing,-Cleaning-Data,-Math)" data-toc-modified-id="-Homework-1-(Topics-1-4:-Data-Structures,-Indexing,-Cleaning-Data,-Math)-1"><span class="toc-item-num">1&nbsp;&nbsp;</span><center> Homework 1 (Topics 1-4: Data Structures, Indexing, Cleaning Data, Math)</center></a></span><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Task-1." data-toc-modified-id="Task-1.-1.0.0.1"><span class="toc-item-num">1.0.0.1&nbsp;&nbsp;</span>Task 1.</a></span></li><li><span><a href="#Варианты-ответов:" data-toc-modified-id="Варианты-ответов:-1.0.0.2"><span class="toc-item-num">1.0.0.2&nbsp;&nbsp;</span>Варианты ответов:</a></span></li><li><span><a href="#Task-2." data-toc-modified-id="Task-2.-1.0.0.3"><span class="toc-item-num">1.0.0.3&nbsp;&nbsp;</span>Task 2.</a></span></li><li><span><a href="#Вопросы:" data-toc-modified-id="Вопросы:-1.0.0.4"><span class="toc-item-num">1.0.0.4&nbsp;&nbsp;</span>Вопросы:</a></span></li><li><span><a href="#Task-3." data-toc-modified-id="Task-3.-1.0.0.5"><span class="toc-item-num">1.0.0.5&nbsp;&nbsp;</span>Task 3.</a></span></li><li><span><a href="#Выберите-все-верные-пункты:" data-toc-modified-id="Выберите-все-верные-пункты:-1.0.0.6"><span class="toc-item-num">1.0.0.6&nbsp;&nbsp;</span>Выберите все верные пункты:</a></span></li><li><span><a href="#Task-4." data-toc-modified-id="Task-4.-1.0.0.7"><span class="toc-item-num">1.0.0.7&nbsp;&nbsp;</span>Task 4.</a></span></li><li><span><a href="#Пояснения:" data-toc-modified-id="Пояснения:-1.0.0.8"><span class="toc-item-num">1.0.0.8&nbsp;&nbsp;</span>Пояснения:</a></span></li></ul></li><li><span><a href="#Информация-для-последующих-заданий" data-toc-modified-id="Информация-для-последующих-заданий-1.0.1"><span class="toc-item-num">1.0.1&nbsp;&nbsp;</span>Информация для последующих заданий</a></span></li></ul></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание данных</a></span><ul class="toc-item"><li><span><a href="#Таблица-transactions.csv" data-toc-modified-id="Таблица-transactions.csv-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>Таблица <code>transactions.csv</code></a></span></li><li><span><a href="#Таблица-gender_train.csv" data-toc-modified-id="Таблица-gender_train.csv-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Таблица <code>gender_train.csv</code></a></span></li><li><span><a href="#Таблица-tr_mcc_codes.csv" data-toc-modified-id="Таблица-tr_mcc_codes.csv-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span>Таблица <code>tr_mcc_codes.csv</code></a></span></li><li><span><a href="#Таблица-tr_types.csv" data-toc-modified-id="Таблица-tr_types.csv-1.1.4"><span class="toc-item-num">1.1.4&nbsp;&nbsp;</span>Таблица <code>tr_types.csv</code></a></span><ul class="toc-item"><li><span><a href="#Task-5." data-toc-modified-id="Task-5.-1.1.4.1"><span class="toc-item-num">1.1.4.1&nbsp;&nbsp;</span>Task 5.</a></span></li><li><span><a href="#Task-6." data-toc-modified-id="Task-6.-1.1.4.2"><span class="toc-item-num">1.1.4.2&nbsp;&nbsp;</span>Task 6.</a></span></li><li><span><a href="#Выберите-все-верные-пункты:" data-toc-modified-id="Выберите-все-верные-пункты:-1.1.4.3"><span class="toc-item-num">1.1.4.3&nbsp;&nbsp;</span>Выберите все верные пункты:</a></span></li><li><span><a href="#Task-7." data-toc-modified-id="Task-7.-1.1.4.4"><span class="toc-item-num">1.1.4.4&nbsp;&nbsp;</span>Task 7.</a></span></li><li><span><a href="#Выберите-все-верные-пункты:" data-toc-modified-id="Выберите-все-верные-пункты:-1.1.4.5"><span class="toc-item-num">1.1.4.5&nbsp;&nbsp;</span>Выберите все верные пункты:</a></span></li><li><span><a href="#Task-8." data-toc-modified-id="Task-8.-1.1.4.6"><span class="toc-item-num">1.1.4.6&nbsp;&nbsp;</span>Task 8.</a></span></li></ul></li></ul></li></ul></li></ul></div>

# <center> Homework 1 (Topics 1-4: Data Structures, Indexing, Cleaning Data, Math)

In [3]:
import pandas as pd
import numpy as np

#### Task 1. 
Задание:

1. Создайте Series из последовательности 15 значений, равномерно разбивающих отрезок [0, 20] (воспользуйтесь функцией linspace)
2. Определите отношение элементов полученной серии к их предыдущим элементам (\*). 
3. В результате необходимо получить среднее полученного вектора, оставив в нём только те значения, которые не более чем 1.5 (\**). 

Выберите из ответов тот, который максимально близок к полученному (c точки зрения абсолютной разницы).

#### Варианты ответов:
    - 1) 1.24
    - 2) 1.18 
    - 3) 0.71
    - 4) 1.13


##### Пояснения:
(\*) Если было бы необходимо найти последовательность из 3-х значений,  
равномерно разбивающих отрезок [0,1], то это были бы значения [0, 0.5, 1]. 

(\**) Если был бы дан список элементов a = [1,2,3,12], отношения элементов к предыдущим будут равны [NaN, 2, 1.5, 4].  
А на последнем этапе в таком примере останется только [1.5] и среднее значение будет также 1.5.

In [4]:

a = pd.Series(np.linspace(0, 20, 15)) 
a = a.pct_change()+1
a = a[a <= 1.5 ].mean() 
a

1.181677812927813

2) 1.18

#### Task 2. 
Выберите все верные ответы касательно следующих 3-х Series:
    - pd.Series('abcde'); (1) 
    - pd.Series(['abcde']); (2) 
    - pd.Series(list('abcde')); (3)
    - pd.Series("abcde"); (4)
    
##### Пояснения:
(\*)
- функция list: в строке каждый символ - это отдельный элемент для list
- квадратные скобки: в квадратных скобках списку передается множество элементов по отдельности через запятую


#### Вопросы:
    - 1) Серия (1) совпадает с серией (2), так как в каждом из случаев серия создаётся из списка строк
    - 2) Серия (2) совпадает с серией (3), так как в каждом из случаев серия создаётся из списка символов 
    - 3) Серия (1) не совпадает с серией (4), так как в (4) используются двойные кавычки "" вместо одинарных '' 

Правильный ответ:  1

#### Task 3. 

По клиенту получены зашумленные данные (объект s типа Series) по его транзакциям. 

Для заданного ниже объекта s проделайте следующее:
1. Создайте новый Series, значения которого совпадают со значениями s, а индексы - целочисленные значения от 2 до 12, не включая 12. 
2. Выберите из s элементы с индексами 3 и 5, после чего просуммируйте их, сохранив результат (1). 
3. Выберите из s только целочисленные элементы и вычислите их дисперсию (2). (\*)

Все полученные результаты округлите до 2-х знаков после запятой.

#### Выберите все верные пункты:
    - 1) Ответ (1) - 642.52
    - 2) Ответ (1) - 91.78
    - 3) Ответ (1) - не может быть определён
    - 4) Ответ (2) - 57591.19
    - 5) Ответ (2) - 210.12
    - 6) Ответ (2) - не может быть определён

##### Пояснения:
(\*) Целочисленные значения - значения, имеющие тип int.
- Дисперсия рассчитывается с помощью функции из библиотеки numpy: np.var( , ddof=0) или встроенной в python функции: .var(ddof=1)

In [5]:
s = pd.Series(data=['1', 2, 3.1, 'hi!', 5, -512, 12.42, 'my_name', 10.10, 98], 
              index=range(6, 26, 2))
new =  pd.Series(list(s), range(2, 12))
new

2           1
3           2
4         3.1
5         hi!
6           5
7        -512
8       12.42
9     my_name
10       10.1
11         98
dtype: object

In [6]:
new[3]+new[5]

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [ ]:
new[new.apply(lambda x: isinstance(x, int))].var(ddof=0)

Ответы: 3,4

#### Task 4. 
1. Сгенерируйте Series из 100 значений нормально распределённой СВ (np.random.normal с дефолтными параметрами - нулевым средним и единичной дисперсией).
2. Возведите каждое значение серии в 3 степень, а значения индекса увеличьте в 3 раза. 
3. Ответьте на следующие вопросы через запятую (без пробелов) (\*)
    - 1) Выведите сумму элементов, строго меньших 2.6, имеющих нечётные значения индекса. 
    - 2) Выведите количество значений серии меньше нуля.

#### Пояснения:
(\*) Если получились ответы 3, 4.32, то в форму необходимо внести их в виде "3,4.32". То есть вещественные числа необходимо разделять точками. Не забудьте про фиксированный seed (его менять не нужно)!
 - Определенное значение seed нужно, чтобы ответы у всех выполняющих
это задание были одинаковые и их можно было проверить (так как генерируются
одинаковые series).
 - Следует внимательнее использовать [ ] для выбора данных по нескольким условиям:
либо выбирать данные последовательно, либо сразу по нескольким условиям,
но через оператор &. Отличие оператора and от оператора &: and - выводит
последнее проверенное значение, & - выводит пересечение значений.
Пример: s[ _  &  _ ].sum() 

In [17]:
np.random.seed(242)
new = pd.Series(np.random.normal(0,1,100))
new **= 3
new.index *= 3

In [18]:
new

0     -0.045698
3      0.003271
6      0.980718
9      6.219024
12    -0.411874
         ...   
285    0.858176
288    0.400829
291   -0.405199
294   -2.606764
297    0.000143
Length: 100, dtype: float64

In [19]:
new[new<0].size

51

In [20]:
new[(new<2.6) & (new.index %2  == 1)].sum()

-15.150748101821666

51, -15.150748101821666

### Информация для последующих заданий
- Для всех последующих заданий будем использовать обезличенные транзакционные банковские данные. Для этого считайте в переменные tr_mcc_codes, tr_types, transactions и gender_train из одноимённых таблиц из папки data. Для таблицы transactions используйте только первые n=1000000 строк. Обратите внимание на разделители внутри каждого из файлов - они могут различаться!

In [8]:
transactions = pd.read_csv("data/transactions.csv")
tr_mcc_codes = pd.read_csv("data/tr_mcc_codes.csv", sep=";")
tr_types = pd.read_csv("data/tr_types.csv", sep=";")
gender_train = pd.read_csv("data/gender_train.csv", sep=";")

In [9]:
dtr_mcc_codes = tr_mcc_codes.head()
dtransactions = transactions.head(1000000)
dtr_types = tr_types.head()
dgender_train = gender_train.head()


## Описание данных
### Таблица ```transactions.csv```
##### Описание
Таблица содержит историю транзакций клиентов банка за один год и три месяца.

##### Формат данных

```
customer_id,tr_datetime,mcc_code,tr_type,amount,term_id
111111,15 01:40:52,1111,1000,-5224,111111
111112,15 15:18:32,3333,2000,-100,11122233
...
```
##### Описание полей

 - ```customer_id``` — идентификатор клиента;
 - ```tr_datetime``` — день и время совершения транзакции (дни нумеруются с начала данных);
 - ```mcc_code``` — mcc-код транзакции;
 - ```tr_type``` — тип транзакции;
 - ```amount``` — сумма транзакции в условных единицах со знаком; ```+``` — начисление средств клиенту (приходная транзакция), ```-``` — списание средств (расходная транзакция);
 - ```term_id``` — идентификатор терминала;


### Таблица ```gender_train.csv```

##### Описание
Данная таблица содержит информацию по полу для части клиентов, для которых он известен. Для остальных клиентов пол неизвестен.

##### Формат данных
```
customer_id,gender
111111,0
111112,1
...
```

##### Описание полей
 - ```customer_id``` — идентификатор клиента;
 - ```gender``` — пол клиента;

### Таблица ```tr_mcc_codes.csv```

##### Описание
Данная таблица содержит описание mcc-кодов транзакций.

##### Формат данных
```
mcc_code;mcc_description
1000;словесное описание mcc-кода 1000
2000;словесное описание mcc-кода 2000
...
```

##### Описание полей
 - ```mcc_code``` – mcc-код транзакции;
 - ```mcc_description``` — описание mcc-кода транзакции.

### Таблица ```tr_types.csv```

##### Описание
Данная таблица содержит описание типов транзакций.

##### Формат данных
```
tr_type;tr_description
1000;словесное описание типа транзакции 1000
2000;словесное описание типа транзакции 2000
...
```

##### Описание полей
 - ```tr_type``` – тип транзакции;
 - ```tr_description``` — описание типа транзакции;

#### Task 5.
1. В tr_types выберите произвольные 100 строк с помощью метода sample (указав при этом random_seed равный 242)
2. В полученной на предыдущем этапе подвыборке найдите долю наблюдений (стобец tr_description), в которой содержится подстрока 'плата' (в любом регистре). (\*)

Выведите ответ в виде вещественного числа, округлённого до двух знаков после запятой, отделив дробную часть точкой в формате "123.45"

##### Пояснения:
(\*) Строки "ПлатА за аренду", "ПлатАза аренду", "ПЛАТА" удовлетворяют условию, так как будучи переведёнными в нижний регистр содержат подстроку "плата".

In [10]:
np.random.seed(242)
d = tr_types.sample(n=100)
d

tr_type                                     tr_description
37      6210                    Возврат покупки.POS Зарубеж. ТУ
50      7034  Перевод на карту/ с карты через АТМ (без взима...
145     2431  Перевод с карты на счет физ.лица в другом банк...
4       4001  Плата за получение наличных в Сбербанке (в дру...
69      8000                   Поправки по счету клиен. Зачисл.
..       ...                                                ...
24      4100                Плата за получение наличных. Россия
130     2371  Списание с карты на карту по операции <перевод...
140     2421  Перевод с карты на счет физ.лица в другом тер....
6       4011  Плата за получение наличных в АТМ Сбербанка (в...
52      7040  Перевод на карту (с карты) через POS (в предел...

[100 rows x 2 columns]

In [11]:
d.tr_description = d.tr_description.apply(lambda x: x.lower())
t = d.tr_description.apply(lambda x: x.find("плата") != -1)
cost = d[t]
cost

tr_type                                     tr_description
4      4001  плата за получение наличных в сбербанке (в дру...
5      4010        плата за получение наличных в атм сбербанка
16     4060  плата за перевод на карту (с карты) через pos ...
30     4500                             плата за превыш.лимита
82     8230  плата за предоставление по запросу клиента док...
81     8220  плата за обработку счета банковской карты клие...
23     4083           оплата услуг банка по поручению автомат.
7      4020  плата за получение наличных через pos в сбербанке
8      4021  плата за получение наличных через pos в сберба...
1      3210  плата за предоставление отчета по счету карты ...
15     4052  плата за перевод на карту (с карты) через атм ...
80     8210                    плата за приост. действия карты
10     4031  плата за взнос наличных через атм (в другом те...
0      3200  плата за предоставление услуг посредством моби...
2      3800  плата за обслуживание банковской карты (за пер...
20     4080                       оплата услуг банка через всп
14     4051  плата за перевод на карту (с карты) через атм ...
26     4200              плата за получение наличных. зарубеж.
17     4061  плата за перевод на карту (с карты) через pos ...
9      4030  плата за взнос наличных через атм (в своем тер...
29     4440  плата за получение через атм выписки о 10 посл...
19     4071  плата за перевод на карту (с карты) через моби...
18     4062  плата за перевод на карту (с карты) через pos ...
12     4041  плата за взнос наличных через pos (в другом те...
24     4100                плата за получение наличных. россия
6      4011  плата за получение наличных в атм сбербанка (в...

In [12]:
part = cost.shape[0] / d.shape[0]
part

0.26

0.26

#### Task 6. 
1. Для поля tr_type датафрейма transactions посчитайте частоту встречаемости всех типов транзакций tr_type в transactions.
2. Из перечисленных вариантов выберите те, которые попали в топ-5 транзакций по частоте встречаемости.

#### Выберите все верные пункты:
    - 1) Выдача наличных в АТМ Сбербанк России
    - 2) Комиссия за обслуживание ссудного счета
    - 3) Списание по требованию	
    - 4) Оплата услуги. Банкоматы СБ РФ
    - 5) Погашение кредита (в пределах одного филиала)
    - 6) Покупка. POS ТУ СБ РФ

In [13]:
type_counts = dtransactions.tr_type.value_counts()
type_counts = type_counts.to_frame()
type_counts = type_counts.rename(columns={"tr_type" : "count"})
type_counts.index.name = "tr_type"
type_counts

count
tr_type        
1010     231117
2010     151166
7070     149006
1110     137658
1030     118975
...         ...
4061          4
4096          4
8146          3
8100          2
1510          2

[72 rows x 1 columns]

In [14]:
dtr_types=tr_types.set_index("tr_type")
dtr_types

tr_description
tr_type                                                   
3200     Плата за предоставление услуг посредством моби...
3210     Плата за предоставление отчета по счету карты ...
3800     Плата за обслуживание банковской карты (за пер...
4000               Плата за получение наличных в Сбербанке
4001     Плата за получение наличных в Сбербанке (в дру...
...                                                    ...
2990                                Списание по требованию
2992     Списание средств для погашения задолженности п...
3001               Комиссия за обслуживание ссудного счета
3030     Плата за обслуживание банковской карты (за пос...
3120                   Плата за досрочный перевыпуск карты

[155 rows x 1 columns]

In [15]:
pd.merge(type_counts, dtr_types, left_on='tr_type', right_on='tr_type',  how='inner')

count                                     tr_description
tr_type                                                           
1010     231117                              Покупка. POS ТУ СБ РФ
2010     151166              Выдача наличных в АТМ Сбербанк России
7070     149006  Перевод на карту (с карты) через Мобильный бан...
1110     137658                             Покупка. POS ТУ Россия
1030     118975                     Оплата услуги. Банкоматы СБ РФ
...         ...                                                ...
2341          4  Списание с карты по операции “перевода с карты...
4061          4  Плата за перевод на карту (с карты) через POS ...
8146          3                                                н/д
8100          2     Списание после проведения претензионной работы
1510          2                                                н/д

[67 rows x 2 columns]

6 1 4

#### Task 7.
1. В датафрейме transactions задайте столбец  customer_id в качестве индекса.
2. Выделите клиента с максимальной суммой транзакции (то есть с максимальным приходом на карту). (\*)
3. Найдите у него наиболее часто встречающийся модуль суммы приходов/расходов. (\**)


#### Выберите все верные пункты:
    - 1) 1122957.89
    - 2) 15721.41
    - 3) 22459.16
    - 4) 13475494.63
    - 5) 107407.78
    - 6) 65019.26

##### Пояснения:
(\*) Если у клиента были транзакции [-10000, 10, 0, -10], то максимумом будет являться значение 10. <br>
(\**) Если у клиента были транзакции [-10000, 10, 0, -10], то наиболее встречающийся модуль суммы транзакций равен 10, и встретился он 2 раза.

In [21]:


from scipy.stats import mode
### Type your code here
data_trans=dtransactions.set_index("customer_id")
data_trans_am = data_trans[['amount']]
gb = data_trans_am.groupby(['customer_id'])
gb = gb['amount'].unique()
gb = gb.to_frame()
gb['max'] = gb['amount'].apply(lambda x: x.max())
gb['mod_mode'] = gb['amount'].apply(lambda x: mode(np.abs(x))[0][0])
gb.sort_values(by='max', ascending=False).head()



amount           max  \
customer_id                                                                    
26782922     [-31442.82, -11935.25, -11229.58, -67377.47, -...  1.459845e+08   
19950741     [-19921.27, -108949.37, -168443.68, -161537.49...  1.347549e+08   
40917212     [-3795.6, -28087.42, -17383.39, -33688.74, -14...  6.737747e+07   
82581973     [-8961.2, -11229.58, -3503.63, -5495.53, -6092...  6.737747e+07   
77695202     [-42223.22, -4522943.99, -1583032.83, 1572141....  6.123489e+07   

             mod_mode  
customer_id            
26782922      2245.92  
19950741     22459.16  
40917212       118.58  
82581973      5614.79  
77695202     67377.47

In [22]:
data_trans
data_trans.sort_values(by='amount', ascending=False).head()

tr_datetime  mcc_code  tr_type        amount term_id
customer_id                                                       
26782922     151 11:35:44      6010     7070  1.459845e+08     NaN
19950741     143 14:08:24      6010     7070  1.347549e+08     NaN
82581973     192 18:18:17      6010     7020  6.737747e+07     NaN
40917212      43 17:10:40      6010     7070  6.737747e+07     NaN
77695202     132 14:07:18      6010     7070  6.123489e+07     NaN

#### Task 8.
1. Найдите максимальную разницу между медианами суммы транзакций, посчитанными при заданных ниже условиях по полю amount из таблицы transactions (\*):
    - Медиана суммы транзакций
    - Медиана суммы транзакций по тем строкам, которые ни в одном из своих столбцов не содержат пустые значения
    - Медиана суммы транзакций по строкам, отсортированным по полю amount в порядке возрастания, и из которых удалены дублирующиеся по столбцам [mcc_code, tr_type] строки, причём при удалении соответстующих дублей остаются только последние из дублирующихся строк (keep='last')

Выведите ответ в виде вещественного числа, округлённого до двух знаков после запятой, отделив дробную часть точкой в формате "123.45"

##### Пояснения:
(\*) Для вычисления максимальной разницы между значениями списка можно использовать функцию np.ptp  
(\**) Если в результате получились значения [1,3,5], то макимальная разница между ними 4 == 5-1.

In [24]:
from statistics import median
data_trans=dtransactions.set_index("customer_id")
data_trans_am = data_trans[['amount']]
gb = data_trans_am.groupby(['customer_id'])
gb = gb['amount'].unique()
gb = gb.to_frame()
gb['median'] = gb['amount'].apply(lambda x: median(x))
np.ptp(gb['median'])


731045.58

In [25]:
data_trans=dtransactions.set_index("customer_id")
data_trans=data_trans.dropna()
data_trans_am = data_trans[['amount']]
gb = data_trans_am.groupby(['customer_id'])
gb = gb['amount'].unique()
gb = gb.to_frame()
gb['median'] = gb['amount'].apply(lambda x: median(x))
np.ptp(gb['median'])

1009539.135

In [26]:
data_trans=dtransactions.set_index("customer_id")
data_trans=data_trans.drop_duplicates(subset=['mcc_code', 'tr_type'], keep='last')
data_trans=data_trans.sort_values(by='amount', ascending=True)
data_trans

tr_datetime  mcc_code  tr_type      amount   term_id
customer_id                                                       
85273426     375 00:00:00      8062     1110 -3833778.22  77047223
14264900     146 12:15:12      4829     2340 -3368873.66       NaN
30627668     318 00:00:00      8299     1200 -3011268.84  00000002
99717689     379 14:03:48      4722     1010 -2078595.05  10306231
36837532     258 00:00:00      5970     1210 -2048585.79  CC100201
...                   ...       ...      ...         ...       ...
37633014     448 13:30:42      6010     7021   622567.85    878293
17389182     126 00:00:00      3501     6100   657379.55       NaN
50762315     437 00:00:00      6012     6210   671665.14  I0110SWG
36734991     412 00:00:00      7011     6100   898366.31    874846
9945756      312 15:31:36      6010     7034   898366.31    190280

[839 rows x 5 columns]

In [27]:
data_trans_am = data_trans[['amount']]
gb = data_trans_am.groupby(['customer_id'])
gb = gb['amount'].unique()
gb = gb.to_frame()
gb['median'] = gb['amount'].apply(lambda x: median(x))
np.ptp(gb['median'])

3945624.83

731045.58, 1009539.135, 3945624.83